### ISSUE TITLE
Multiple tile sources in a single leafletLayer

### ISSUE THREAD

**Issue by bdon** [protomaps/protomaps.js#33]
* Label rules and paint rules optionally specify the data (TileCache) they fetch from
* should use Promise.allSettled to wait for all tiles to be fetched
* What should the behavior be when one of the tiles fail?
* Fetches to multiple sources should happen in parallel; should drawing block on having all data? Or should we start drawing, and each individual layer awaits the data availability? We need to work around the constraint that drawing is always back-to-front.

**Comment by bdon** [comment]
Brain dump about branch `tile-smushing`

Overall approach:
* Block on all tile promises being settled before painting anything
* Internal rules state isn't mutable, to change sources pass new Rules
* Default `dataSource` has empty string as name, to avoid having to explicitly name It in every Rule in the general case

Progress:
* Painter can paint from multiple dataSources, even when their overzooming does not match

Challenges:
* Static map clipping and X-ray broken still
* Labeling is currently keyed by dataTile, but for a given zoom level the dataTile may not be consistent among data sources because of different levelDiffs or different overzooming states. Labeling keys and invalidation needs to be more sophisticated to bookkeep on a key specific to the individual data source

**Comment by bdon** [comment]
current spike hits major layout/paint jankiness around zoom 10

**Comment by bdon** [comment]
layout at massive overzooms for data tiles is taking 60ms+

**Comment by bdon** [comment]
http://localhost:8000/examples/leaflet.html#16/41.1594/-120.7284 kills the browser, probably due to layout above^

**Comment by bdon** [comment]
layer-level options that need to be moved to source-level: `levelDiff`, `maxDataZoom`

**Comment by bdon** [comment]
```json
{
  "":{source:"https://example.com/{z}/{x}/{y}.pbf",levelDiff:2,maxDataZoom:14}
  "other":{source:"https://example.com/foo.pmtiles,levelDiff:2,maxDataZoom:8}
}
```

**Comment by bdon** [comment]
Possible bug related to labels across tile edges: but not easy to reproduce; still investigating

**Comment by bdon** [comment]
Implemented in 1.19
